File: test/java/brs/services/impl/AliasServiceImplTest.java
package brs.services.impl;

import brs.Alias;
import brs.Signum;
import brs.Alias.Offer;
import brs.Attachment.MessagingAliasAssignment;
import brs.Attachment.MessagingAliasSell;
import brs.Transaction;
import brs.common.AbstractUnitTest;
import brs.common.QuickMocker;
import brs.db.BurstKey;
import brs.db.BurstKey.LongKeyFactory;
import brs.db.VersionedEntityTable;
import brs.db.store.AliasStore;
import brs.fluxcapacitor.FluxCapacitor;
import brs.fluxcapacitor.FluxValues;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.powermock.core.classloader.annotations.PrepareForTest;
import org.powermock.modules.junit4.PowerMockRunner;

import java.util.Collection;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.*;
import static org.powermock.api.mockito.PowerMockito.mock;
import static org.powermock.api.mockito.PowerMockito.mockStatic;

@RunWith(PowerMockRunner.class)
@PrepareForTest(Signum.class)
public class AliasServiceImplTest extends AbstractUnitTest {

  private AliasServiceImpl t;

  private AliasStore aliasStoreMock;
  private VersionedEntityTable<Alias> aliasTableMock;
  private BurstKey.LongKeyFactory<Alias> aliasDbKeyFactoryMock;
  private VersionedEntityTable<Offer> offerTableMock;
  private BurstKey.LongKeyFactory<Offer> offerDbKeyFactoryMock;

  @Before
  public void setUp() {
    mockStatic(Signum.class);

    aliasStoreMock = mock(AliasStore.class);
    aliasTableMock = mock(VersionedEntityTable.class);
    aliasDbKeyFactoryMock = mock(LongKeyFactory.class);
    offerTableMock = mock(VersionedEntityTable.class);
    offerDbKeyFactoryMock = mock(LongKeyFactory.class);

    FluxCapacitor mockFluxCapacitor = QuickMocker.fluxCapacitorEnabledFunctionalities(FluxValues.PRE_POC2, FluxValues.DIGITAL_GOODS_STORE);
    when(Signum.getFluxCapacitor()).thenReturn(mockFluxCapacitor);

    when(aliasStoreMock.getAliasTable()).thenReturn(aliasTableMock);
    when(aliasStoreMock.getAliasDbKeyFactory()).thenReturn(aliasDbKeyFactoryMock);
    when(aliasStoreMock.getOfferTable()).thenReturn(offerTableMock);
    when(aliasStoreMock.getOfferDbKeyFactory()).thenReturn(offerDbKeyFactoryMock);

    t = new AliasServiceImpl(aliasStoreMock);
  }

  @Test
  public void getAlias() {
    final String aliasName = "aliasName";
    final Alias mockAlias = mock(Alias.class);

    when(aliasStoreMock.getAlias(eq(aliasName), eq(0L))).thenReturn(mockAlias);

    assertEquals(mockAlias, t.getAlias(aliasName, 0L));
  }

  @Test
  public void getAlias_byId() {
    final long id = 123l;
    final BurstKey mockKey = mock(BurstKey.class);
    final Alias mockAlias = mock(Alias.class);

    when(aliasDbKeyFactoryMock.newKey(eq(id))).thenReturn(mockKey);
    when(aliasTableMock.get(eq(mockKey))).thenReturn(mockAlias);

    assertEquals(mockAlias, t.getAlias(id));
  }

  @Test
  public void getOffer() {
    final Long aliasId = 123l;
    final Alias mockAlias = mock(Alias.class);
    when(mockAlias.getId()).thenReturn(aliasId);
    final BurstKey mockOfferKey = mock(BurstKey.class);
    final Offer mockOffer = mock(Offer.class);

    when(offerDbKeyFactoryMock.newKey(eq(aliasId))).thenReturn(mockOfferKey);
    when(offerTableMock.get(eq(mockOfferKey))).thenReturn(mockOffer);

    assertEquals(mockOffer, t.getOffer(mockAlias));
  }

  @Test
  public void getAliasCount() {
    when(aliasTableMock.getCount()).thenReturn(5);
    assertEquals(5L, t.getAliasCount());
  }

  @Test
  public void getAliasesByOwner() {
    final long accountId = 123L;
    final int from = 0;
    final int to = 1;

    final Collection<Alias> mockAliasIterator = mockCollection();

    when(aliasStoreMock.getAliasesByOwner(eq(accountId), isNull(), eq(0L), eq(from), eq(to))).thenReturn(mockAliasIterator);

    assertEquals(mockAliasIterator, t.getAliasesByOwner(accountId, null, 0L, from, to).getCollection());
  }

  @Test
  public void addOrUpdateAlias_addAlias() {
    final Transaction transaction = mock(Transaction.class);
    when(transaction.getSenderId()).thenReturn(123L);
    when(transaction.getBlockTimestamp()).thenReturn(34);

    final MessagingAliasAssignment attachment = mock(MessagingAliasAssignment.class);
    when(attachment.getAliasURI()).thenReturn("aliasURI");

    t.addOrUpdateAlias(transaction, attachment);

    final ArgumentCaptor<Alias> savedAliasCaptor = ArgumentCaptor.forClass(Alias.class);

    verify(aliasTableMock).insert(savedAliasCaptor.capture());

    final Alias savedAlias = savedAliasCaptor.getValue();
    assertNotNull(savedAlias);

    assertEquals(transaction.getSenderId(), savedAlias.getAccountId());
    assertEquals(transaction.getBlockTimestamp(), savedAlias.getTimestamp());
    assertEquals(attachment.getAliasURI(), savedAlias.getAliasURI());
  }

  @Test
  public void addOrUpdateAlias_updateAlias() {
    final String aliasName = "aliasName";
    final Alias mockAlias = mock(Alias.class);

    when(aliasStoreMock.getAlias(eq(aliasName), eq(0L))).thenReturn(mockAlias);

    final Transaction transaction = mock(Transaction.class);
    when(transaction.getSenderId()).thenReturn(123L);
    when(transaction.getBlockTimestamp()).thenReturn(34);

    final MessagingAliasAssignment attachment = mock(MessagingAliasAssignment.class);
    when(attachment.getAliasName()).thenReturn(aliasName);
    when(attachment.getAliasURI()).thenReturn("aliasURI");

    t.addOrUpdateAlias(transaction, attachment);

    verify(mockAlias).setAccountId(eq(transaction.getSenderId()));
    verify(mockAlias).setTimestamp(eq(transaction.getBlockTimestamp()));
    verify(mockAlias).setAliasURI(eq(attachment.getAliasURI()));

    verify(aliasTableMock).insert(eq(mockAlias));
  }

  @Test
  public void sellAlias_forBurst_newOffer() {
    final String aliasName = "aliasName";
    final long aliasId = 123L;
    final Alias mockAlias = mock(Alias.class);
    when(mockAlias.getId()).thenReturn(aliasId);

    when(aliasStoreMock.getAlias(eq(aliasName), eq(0L))).thenReturn(mockAlias);

    final BurstKey mockOfferKey = mock(BurstKey.class);
    when(offerDbKeyFactoryMock.newKey(eq(aliasId))).thenReturn(mockOfferKey);

    final long priceNQT = 500L;

    final long newOwnerId = 234L;
    final int timestamp = 567;

    final Transaction transaction = mock(Transaction.class);
    final MessagingAliasSell attachment = mock(MessagingAliasSell.class);
    when(attachment.getAliasName()).thenReturn(aliasName);
    when(attachment.getPriceNQT()).thenReturn(priceNQT);
    when(transaction.getBlockTimestamp()).thenReturn(timestamp);
    when(transaction.getRecipientId()).thenReturn(newOwnerId);

    t.sellAlias(transaction, attachment);

    ArgumentCaptor<Offer> mockOfferCaptor = ArgumentCaptor.forClass(Offer.class);

    verify(offerTableMock).insert(mockOfferCaptor.capture());

    final Offer savedOffer = mockOfferCaptor.getValue();
    assertEquals(newOwnerId, savedOffer.getBuyerId());
    assertEquals(priceNQT, savedOffer.getPriceNQT());
  }

  @Test
  public void sellAlias_forBurst_offerExists() {
    final String aliasName = "aliasName";
    final long aliasId = 123L;
    final Alias mockAlias = mock(Alias.class);
    when(mockAlias.getId()).thenReturn(aliasId);

    when(aliasStoreMock.getAlias(eq(aliasName), eq(0L))).thenReturn(mockAlias);

    final BurstKey mockOfferKey = mock(BurstKey.class);
    final Offer mockOffer = mock(Offer.class);
    when(offerDbKeyFactoryMock.newKey(eq(aliasId))).thenReturn(mockOfferKey);
    when(offerTableMock.get(eq(mockOfferKey))).thenReturn(mockOffer);

    final long priceNQT = 500L;

    final long newOwnerId = 234L;
    final int timestamp = 567;

    final Transaction transaction = mock(Transaction.class);
    final MessagingAliasSell attachment = mock(MessagingAliasSell.class);
    when(attachment.getAliasName()).thenReturn(aliasName);
    when(attachment.getPriceNQT()).thenReturn(priceNQT);
    when(transaction.getBlockTimestamp()).thenReturn(timestamp);
    when(transaction.getRecipientId()).thenReturn(newOwnerId);

    t.sellAlias(transaction, attachment);

    verify(mockOffer).setPriceNQT(eq(priceNQT));
    verify(mockOffer).setBuyerId(eq(newOwnerId));

    verify(offerTableMock).insert(eq(mockOffer));
  }

  @Test
  public void sellAlias_forFree() {
    final String aliasName = "aliasName";
    final long aliasId = 123L;
    final Alias mockAlias = mock(Alias.class);
    when(mockAlias.getId()).thenReturn(aliasId);

    when(aliasStoreMock.getAlias(eq(aliasName), eq(0L))).thenReturn(mockAlias);

    final BurstKey mockOfferKey = mock(BurstKey.class);
    final Offer mockOffer = mock(Offer.class);
    when(offerDbKeyFactoryMock.newKey(eq(aliasId))).thenReturn(mockOfferKey);
    when(offerTableMock.get(eq(mockOfferKey))).thenReturn(mockOffer);

    final long priceNQT = 0L;

    final long newOwnerId = 234L;
    final int timestamp = 567;

    final Transaction transaction = mock(Transaction.class);
    final MessagingAliasSell attachment = mock(MessagingAliasSell.class);
    when(attachment.getAliasName()).thenReturn(aliasName);
    when(attachment.getPriceNQT()).thenReturn(priceNQT);
    when(transaction.getBlockTimestamp()).thenReturn(timestamp);
    when(transaction.getRecipientId()).thenReturn(newOwnerId);

    t.sellAlias(transaction, attachment);

    verify(mockAlias).setAccountId(newOwnerId);
    verify(mockAlias).setTimestamp(eq(timestamp));
    verify(aliasTableMock).insert(mockAlias);

    verify(offerTableMock).delete(eq(mockOffer));
  }

  @Test
  public void changeOwner() {
    final String aliasName = "aliasName";
    final long aliasId = 123L;
    final Alias mockAlias = mock(Alias.class);
    when(mockAlias.getId()).thenReturn(aliasId);

    when(aliasStoreMock.getAlias(eq(aliasName), eq(0L))).thenReturn(mockAlias);

    final BurstKey mockOfferKey = mock(BurstKey.class);
    final Offer mockOffer = mock(Offer.class);
    when(offerDbKeyFactoryMock.newKey(eq(aliasId))).thenReturn(mockOfferKey);
    when(offerTableMock.get(eq(mockOfferKey))).thenReturn(mockOffer);

    final long newOwnerId = 234L;
    final int timestamp = 567;

    t.changeOwner(newOwnerId, mockAlias, timestamp, true);

    verify(mockAlias).setAccountId(newOwnerId);
    verify(mockAlias).setTimestamp(eq(timestamp));
    verify(aliasTableMock).insert(mockAlias);

    verify(offerTableMock).delete(eq(mockOffer));
  }
}

File: test/java/brs/services/impl/AliasServiceImplTest.kt
package brs.services.impl

import brs.Alias
import brs.Alias.Offer
import brs.Attachment.MessagingAliasAssignment
import brs.Attachment.MessagingAliasSell
import brs.Transaction
import brs.common.AbstractUnitTest
import brs.db.BurstKey
import brs.db.BurstKey.LongKeyFactory
import brs.db.VersionedEntityTable
import brs.db.store.AliasStore
import com.nhaarman.mockitokotlin2.*
import org.junit.Before
import org.junit.Test

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull

class AliasServiceImplTest : AbstractUnitTest() {

    private var t: AliasServiceImpl? = null

    private var aliasStoreMock: AliasStore? = null
    private var aliasTableMock: VersionedEntityTable<Alias>? = null
    private var aliasDbKeyFactoryMock: BurstKey.LongKeyFactory<Alias>? = null
    private var offerTableMock: VersionedEntityTable<Offer>? = null
    private var offerDbKeyFactoryMock: BurstKey.LongKeyFactory<Offer>? = null

    @Before
    fun setUp() {
        aliasStoreMock = mock()
        aliasTableMock = mock()
        aliasDbKeyFactoryMock = mock()
        offerTableMock = mock()
        offerDbKeyFactoryMock = mock()

        whenever(aliasStoreMock!!.aliasTable).thenReturn(aliasTableMock)
        whenever(aliasStoreMock!!.aliasDbKeyFactory).thenReturn(aliasDbKeyFactoryMock)
        whenever(aliasStoreMock!!.offerTable).thenReturn(offerTableMock)
        whenever(aliasStoreMock!!.offerDbKeyFactory).thenReturn(offerDbKeyFactoryMock)

        t = AliasServiceImpl(aliasStoreMock!!)
    }

    @Test
    fun getAlias() {
        val aliasName = "aliasName"
        val mockAlias = mock<Alias>()

        whenever(aliasStoreMock!!.getAlias(eq(aliasName))).thenReturn(mockAlias)

        assertEquals(mockAlias, t!!.getAlias(aliasName))
    }

    @Test
    fun getAlias_byId() {
        val id = 123L
        val mockKey = mock<BurstKey>()
        val mockAlias = mock<Alias>()

        whenever(aliasDbKeyFactoryMock!!.newKey(eq(id))).thenReturn(mockKey)
        whenever(aliasTableMock!!.get(eq(mockKey))).thenReturn(mockAlias)

        assertEquals(mockAlias, t!!.getAlias(id))
    }

    @Test
    fun getOffer() {
        val aliasId = 123L
        val mockAlias = mock<Alias>()
        whenever(mockAlias.id).thenReturn(aliasId)
        val mockOfferKey = mock<BurstKey>()
        val mockOffer = mock<Offer>()

        whenever(offerDbKeyFactoryMock!!.newKey(eq<Long>(aliasId))).thenReturn(mockOfferKey)
        whenever(offerTableMock!!.get(eq(mockOfferKey))).thenReturn(mockOffer)

        assertEquals(mockOffer, t!!.getOffer(mockAlias))
    }

    @Test
    fun getAliasCount() {
        whenever(aliasTableMock!!.count).thenReturn(5)
        assertEquals(5L, t!!.aliasCount)
    }

    @Test
    fun getAliasesByOwner() {
        val accountId = 123L
        val from = 0
        val to = 1

        val mockAliasIterator = mockCollection<Alias>()

        whenever(aliasStoreMock!!.getAliasesByOwner(eq(accountId), eq(from), eq(to))).thenReturn(mockAliasIterator)

        assertEquals(mockAliasIterator, t!!.getAliasesByOwner(accountId, from, to))
    }

    @Test
    fun addOrUpdateAlias_addAlias() {
        val transaction = mock<Transaction>()
        whenever(transaction.senderId).thenReturn(123L)
        whenever(transaction.blockTimestamp).thenReturn(34)

        val attachment = mock<MessagingAliasAssignment>()
        whenever(attachment.aliasURI).thenReturn("aliasURI")

        t!!.addOrUpdateAlias(transaction, attachment)

        val savedAliasCaptor = argumentCaptor<Alias>()

        verify(aliasTableMock!!).insert(savedAliasCaptor.capture())

        val savedAlias = savedAliasCaptor.firstValue
        assertNotNull(savedAlias)

        assertEquals(transaction.senderId, savedAlias.accountId)
        assertEquals(transaction.blockTimestamp.toLong(), savedAlias.timestamp.toLong())
        assertEquals(attachment.aliasURI, savedAlias.aliasURI)
    }

    @Test
    fun addOrUpdateAlias_updateAlias() {
        val aliasName = "aliasName"
        val mockAlias = mock<Alias>()

        whenever(aliasStoreMock!!.getAlias(eq(aliasName))).thenReturn(mockAlias)

        val transaction = mock<Transaction>()
        whenever(transaction.senderId).thenReturn(123L)
        whenever(transaction.blockTimestamp).thenReturn(34)

        val attachment = mock<MessagingAliasAssignment>()
        whenever(attachment.aliasName).thenReturn(aliasName)
        whenever(attachment.aliasURI).thenReturn("aliasURI")

        t!!.addOrUpdateAlias(transaction, attachment)

        verify(mockAlias).accountId = eq(transaction.senderId)
        verify(mockAlias).timestamp = eq(transaction.blockTimestamp)
        verify(mockAlias).aliasURI = eq(attachment.aliasURI)

        verify(aliasTableMock!!).insert(eq(mockAlias))
    }

    @Test
    fun sellAlias_forBurst_newOffer() {
        val aliasName = "aliasName"
        val aliasId = 123L
        val mockAlias = mock<Alias>()
        whenever(mockAlias.id).thenReturn(aliasId)

        whenever(aliasStoreMock!!.getAlias(eq(aliasName))).thenReturn(mockAlias)

        val mockOfferKey = mock<BurstKey>()
        whenever(offerDbKeyFactoryMock!!.newKey(eq(aliasId))).thenReturn(mockOfferKey)

        val priceNQT = 500L

        val newOwnerId = 234L
        val timestamp = 567

        val transaction = mock<Transaction>()
        val attachment = mock<MessagingAliasSell>()
        whenever(attachment.aliasName).thenReturn(aliasName)
        whenever(attachment.priceNQT).thenReturn(priceNQT)
        whenever(transaction.blockTimestamp).thenReturn(timestamp)
        whenever(transaction.recipientId).thenReturn(newOwnerId)

        t!!.sellAlias(transaction, attachment)

        val mockOfferCaptor = argumentCaptor<Offer>()

        verify(offerTableMock!!).insert(mockOfferCaptor.capture())

        val savedOffer = mockOfferCaptor.firstValue
        assertEquals(newOwnerId, savedOffer.buyerId)
        assertEquals(priceNQT, savedOffer.priceNQT)
    }

    @Test
    fun sellAlias_forBurst_offerExists() {
        val aliasName = "aliasName"
        val aliasId = 123L
        val mockAlias = mock<Alias>()
        whenever(mockAlias.id).thenReturn(aliasId)

        whenever(aliasStoreMock!!.getAlias(eq(aliasName))).thenReturn(mockAlias)

        val mockOfferKey = mock<BurstKey>()
        val mockOffer = mock<Offer>()
        whenever(offerDbKeyFactoryMock!!.newKey(eq(aliasId))).thenReturn(mockOfferKey)
        whenever(offerTableMock!!.get(eq(mockOfferKey))).thenReturn(mockOffer)

        val priceNQT = 500L

        val newOwnerId = 234L
        val timestamp = 567

        val transaction = mock<Transaction>()
        val attachment = mock<MessagingAliasSell>()
        whenever(attachment.aliasName).thenReturn(aliasName)
        whenever(attachment.priceNQT).thenReturn(priceNQT)
        whenever(transaction.blockTimestamp).thenReturn(timestamp)
        whenever(transaction.recipientId).thenReturn(newOwnerId)

        t!!.sellAlias(transaction, attachment)

        verify(mockOffer).priceNQT = eq(priceNQT)
        verify(mockOffer).buyerId = eq(newOwnerId)

        verify(offerTableMock!!).insert(eq(mockOffer))
    }

    @Test
    fun sellAlias_forFree() {
        val aliasName = "aliasName"
        val aliasId = 123L
        val mockAlias = mock<Alias>()
        whenever(mockAlias.id).thenReturn(aliasId)

        whenever(aliasStoreMock!!.getAlias(eq(aliasName))).thenReturn(mockAlias)

        val mockOfferKey = mock<BurstKey>()
        val mockOffer = mock<Offer>()
        whenever(offerDbKeyFactoryMock!!.newKey(eq(aliasId))).thenReturn(mockOfferKey)
        whenever(offerTableMock!!.get(eq(mockOfferKey))).thenReturn(mockOffer)

        val priceNQT = 0L

        val newOwnerId = 234L
        val timestamp = 567

        val transaction = mock<Transaction>()
        val attachment = mock<MessagingAliasSell>()
        whenever(attachment.aliasName).thenReturn(aliasName)
        whenever(attachment.priceNQT).thenReturn(priceNQT)
        whenever(transaction.blockTimestamp).thenReturn(timestamp)
        whenever(transaction.recipientId).thenReturn(newOwnerId)

        t!!.sellAlias(transaction, attachment)

        verify(mockAlias).accountId = newOwnerId
        verify(mockAlias).timestamp = eq(timestamp)
        verify(aliasTableMock!!).insert(mockAlias)

        verify(offerTableMock!!).delete(eq(mockOffer))
    }

    @Test
    fun changeOwner() {
        val aliasName = "aliasName"
        val aliasId = 123L
        val mockAlias = mock<Alias>()
        whenever(mockAlias.id).thenReturn(aliasId)

        whenever(aliasStoreMock!!.getAlias(eq(aliasName))).thenReturn(mockAlias)

        val mockOfferKey = mock<BurstKey>()
        val mockOffer = mock<Offer>()
        whenever(offerDbKeyFactoryMock!!.newKey(eq(aliasId))).thenReturn(mockOfferKey)
        whenever(offerTableMock!!.get(eq(mockOfferKey))).thenReturn(mockOffer)

        val newOwnerId = 234L
        val timestamp = 567

        t!!.changeOwner(newOwnerId, aliasName, timestamp)

        verify(mockAlias).accountId = newOwnerId
        verify(mockAlias).timestamp = eq(timestamp)
        verify(aliasTableMock!!).insert(mockAlias)

        verify(offerTableMock!!).delete(eq(mockOffer))
    }
}
